src/artifacts-cred.py

```diff
@@ -0,0 +1,177 @@
+                    " the .NET SDK and ensure 'dotnet' is in your PATH. Error: "
+                )
+                raise Exception(message + str(e))
+
+            #TODO: fix the path
+            tool_path = os.path.join(
+                os.path.dirname(os.path.abspath(__file__)),
+                "plugins",
+                "plugins",
+                "netcore",
+                "CredentialProvider.Microsoft",
+                "CredentialProvider.Microsoft.dll",
+            )
+            self.exe = [get_runtime_path(), "exec", tool_path]
+
+        if not os.path.exists(tool_path):
+            raise RuntimeError("Unable to find credential provider in the expected path: " + tool_path)
+
+
+    def get_credentials(self, url):
+        # Public feed short circuit: return nothing if the endpoint is public (can authenticate without credentials).
+        if self._can_authenticate(url, None):
+            return None, None
+
+        # Getting credentials with IsRetry=false; the credentials may come from the cache
+        username, password = self._get_credentials_from_credential_provider(url, is_retry=False)
+        
+        # Do not attempt to validate if the credentials could not be obtained
+        if username is None or password is None:
+            return username, password
+
+        # Make sure the credentials are still valid (i.e. not expired)
+        if self._can_authenticate(url, (username, password)):
+            return username, password
+
+        # The cached credentials are expired; get fresh ones with IsRetry=true
+        return self._get_credentials_from_credential_provider(url, is_retry=True)
+
+    def _can_authenticate(self, url, auth):
+        response = requests.get(url, auth=auth)
+
+        return response.status_code < 500 and \
+            response.status_code != 401 and \
+            response.status_code != 403
+
+
+    def _get_credentials_from_credential_provider(self, url, is_retry):
+        non_interactive = self._NON_INTERACTIVE_VAR_NAME in os.environ and \
+            os.environ[self._NON_INTERACTIVE_VAR_NAME] and \
+            str(os.environ[self._NON_INTERACTIVE_VAR_NAME]).lower() == "true"
+
+        proc = Popen(
+            self.exe + [
+                "-Uri", url,
+                "-IsRetry", str(is_retry),
+                "-NonInteractive", str(non_interactive),
+                "-CanShowDialog", "False",
+                "-OutputFormat", "Json"
+            ],
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE
+        )
+
+        # Read all standard error first, which may either display
+        # errors from the credential provider or instructions
+        # from it for Device Flow authentication.
+        for stderr_line in iter(proc.stderr.readline, b''):
+            line = stderr_line.decode("utf-8", "ignore")
+            sys.stdout.write(line)
```

Comment:
This should be written to stderr. We don't want anything but the token going to stdout.